library/bigip_software_facts.py

```diff
@@ -0,0 +1,317 @@
+        hotfixes:
+            - build: "2.0.204"
+              fileSize: "1997 MB",
+              lastModified: "Sun Oct  2 20:50:04 2016",
+              name: "12.1.1-hf2.iso",
+              product: "BIG-IP",
+              version: "12.1.1"   
+volumes:     
+    description:
+        List the volumes present on device.
+    returned: changed
+    type: list of dict
+    sample:
+       volumes:
+            - basebuild: "0.0.184",
+              build: "0.0.184",
+              name: "HD1.2",
+              product: "BIG-IP",
+              status: "complete",
+              version: "12.1.1"              
+'''
+
+
+from ansible.module_utils.f5_utils import (
+    AnsibleF5Client,
+    AnsibleF5Parameters,
+    HAS_F5SDK,
+    F5ModuleError,
+)
+
+
+class Parameters(AnsibleF5Parameters):
+    returnables = [
+        'name', 'build', 'version', 'product',
+        'lastModified', 'basebuild', 'status',
+        'active', 'fileSize'
+        ]
+
+    @property
+    def include(self):
+        requested = self._values['include']
+        valid = ['all', 'volume', 'image', 'hotfix']
+
+        if not set(requested).issubset(set(valid)):
+            raise F5ModuleError(
+                'Include parameter may only be specified as one or more of the following: {0}'.format(', '.join(valid))
+            )
+
+        if 'all' in requested:
+            return ['all']
+        else:
+            return requested
+
+    @property
+    def filter(self):
+        requested = self._values['filter']
+        keys = ['name', 'build', 'version', 'status', 'active']
+        error = '"{0}" is not a valid filter format. Filters must have key:value format'.format(requested)
+        if requested is None:
+            return None
+
+        if ':' not in requested:
+            raise F5ModuleError(error)
+
+        key, value = requested.split(':')
+
+        if len(key) == 0 or len(value) == 0:
+            raise F5ModuleError(error)
+
+        if key not in keys:
+            raise F5ModuleError('"{0}" is not a supported filter. '
+                                'Supported key values are: {1}'.format(key, ', '.join(keys)))
+
+        return key, value
+
+
+class ModuleManager(object):
+    def __init__(self, client):
+        self.client = client
+
+    def exec_module(self):
+        manager = BaseManager(self.client)
+
+        return manager.exec_module()
+
+
+class BaseManager(object):
+    def __init__(self, client):
+        self.client = client
+        self.want = Parameters(self.client.module.params)
+        self.result = dict()
+        self.include = self.want.include
+        self.filter = self.want.filter
+
+    def exec_module(self):
+        result = dict()
+
+        if 'all' in self.include:
+            facts = self.get_all_facts()
+        else:
+            facts = self.get_selected_facts()
+
+        result.update(**facts)
+        result.update(dict(changed=True))
+        return result
+
+    def get_all_facts(self):
```

Comment:
something like this method, but not this method, should live in the module manager. There is no need to list all the managers here. Instead, with the get_managers() method that you'll find in bigip_ssl_certificate, you should be able to call all the get_facts on their own